src/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -168,6 +172,21 @@ public override bool TryComplete(Exception error)
 
                     // Mark that we're done writing.
                     parent._doneWriting = error ?? ChannelUtilities.s_doneWritingSentinel;
+
+                    blockedWriters = parent._blockedWriters;
```

Comment:
How does this help?  blockedWriters and parent._blockedWriters both now refer to the same instance; accessing blockedWriters after the lock is no different than accessing parent._blockedWriters after the lock, which is underlying issue being addressed.

Reply:
I thought we set it to null but it looks not. we always have a valid _blockedWriters. do you think keeping  
                ChannelUtilities.FailInteractors<WriterInteractor<T>, VoidResult>(_blockedWriters, ChannelUtilities.CreateInvalidCompletionException(error));

inside a lock block will be a problem?

Reply:
> inside a lock block will be a problem?

Yes.  These interactors are created with the AllowSynchronousContinuations setting, which translates inversely to RunContinuationsAsynchronously on the TaskCompletionSource used.  If AllowSynchronousContinuations is true, RunContinuationsAsynchronously won't be set, and completing the TCS will cause any continuations from its task to be invoked synchronously (the point of AllowSynchronousContinuations).  If the TCS is completed while we hold a lock, those continuations will be invoked while we hold the lock.  And that leads to badness, because a) invariants are broken while we hold the lock, and b) it can lead to deadlocks.

> and does it has to be called after

It should be.  Not doing so doesn't strictly lead to a problem, but it leads to a strange situation where we wake up a blocked writer stating that the channel has been closed, and there aren't any more items to receive, but if it checks the channel's completion task, it may not yet be set to completed, due to the race condition.  If we instead compete the blocked writers after transitioning the completion task to a completed state, there is no race condition, and they're guaranteed to see that the block is closed.

I apparently broke this when adding some support for AllowSynchronousContinuations to the bounded channel.  The easy "fix" is to remove that support, e.g. by changing this line:
https://github.com/dotnet/corefx/blob/master/src/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs#L106
to pass true for runContinuationsAsynchronously.  But that's not really the "right" fix, as it removes functionality, and we actually need to be going in the opposite direction, adding more support for AllowSynchronousContinuations rather than less.

I won't be able to look at this in more depth until Monday.

Reply:
thanks @stephentoub that clarifies a lot. I don't like removing any feature either so I wouldn't remove AllowSynchronousContinuations support. what I suggest here is I'll keep 

```C#
ChannelUtilities.FailInteractors<WriterInteractor<T>, VoidResult>(parent._blockedWriters, ChannelUtilities.CreateInvalidCompletionException(error));
```

outside the lock and continue with this PR. I'll open issue tracking this issue and I'll try to look for a solution too. on Monday, we can talk about that deeply and come up with the right fix.

I'll not port any changes to preview 1 branch before having the full fix.

Is this ok with you? and sorry interrupting you during your time off :-)

Reply:
> Is this ok with you?

I suggest keeping the while(true) change, keeping the ConfigureAwait change, and keeping the additional comment about the issue, but reverting the rest; that issue can then be addressed fully subsequently.